src/scheduled_tasks.rs

```diff
@@ -323,62 +328,65 @@ fn update_banned_when_expired(conn: &mut PgConnection) {
   info!("Updating instances software and versions...");
 
-  let client = match Client::builder()
+  let client = Client::builder()
     .user_agent(user_agent)
     .timeout(REQWEST_TIMEOUT)
-    .build()
-  {
-    Ok(client) => client,
-    Err(e) => {
-      error!("Failed to build reqwest client: {}", e);
-      return;
-    }
-  };
+    .build()?;
 
-  let instances = match instance::table.get_results::<Instance>(conn) {
-    Ok(instances) => instances,
-    Err(e) => {
-      error!("Failed to get instances: {}", e);
-      return;
-    }
-  };
+  let instances = instance::table.get_results::<Instance>(conn)?;
 
   for instance in instances {
     let node_info_url = format!("https://{}/nodeinfo/2.0.json", instance.domain);
 
-    // Skip it if it can't connect
-    let res = client
-      .get(&node_info_url)
-      .send()
-      .ok()
-      .and_then(|t| t.json::<NodeInfo>().ok());
-
-    if let Some(node_info) = res {
-      let software = node_info.software.as_ref();
-      let form = InstanceForm::builder()
-        .domain(instance.domain)
-        .software(software.and_then(|s| s.name.clone()))
-        .version(software.and_then(|s| s.version.clone()))
-        .updated(Some(naive_now()))
-        .build();
-
-      match diesel::update(instance::table.find(instance.id))
-        .set(form)
-        .execute(conn)
-      {
-        Ok(_) => {
-          info!("Done.");
+    // The `updated` column is used to check if instances are alive. If it is more than three days
+    // in the past, no outgoing activities will be sent to that instance. However not every
+    // Fediverse instance has a valid Nodeinfo endpoint (its not required for Activitypub). That's
+    // why we always need to mark instances as updated if they are alive.
+    let default_form = InstanceForm::builder()
+      .domain(instance.domain.clone())
+      .updated(Some(naive_now()))
+      .build();
+    let form = match client.get(&node_info_url).send() {
+      Ok(res) if res.status().is_client_error() => {
+        // Instance doesnt have nodeinfo but sent a response, consider it alive
+        Some(default_form)
+      }
+      Ok(res) => match res.json::<NodeInfo>() {
+        Ok(node_info) => {
+          // Instance sent valid nodeinfo, write it to db
+          Some(
+            InstanceForm::builder()
+              .domain(instance.domain)
+              .updated(Some(naive_now()))
+              .software(node_info.software.and_then(|s| s.name))
+              .version(node_info.version.clone())
+              .build(),
+          )
         }
-        Err(e) => {
-          error!("Failed to update site instance software: {}", e);
-          return;
+        Err(_) => {
+          // No valid nodeinfo but valid HTTP response, consider instance alive
+          Some(default_form)
         }
+      },
+      Err(_) => {
+        // dead instance, do nothing
+        None
       }
+    };
```

Comment:
This code is quite confusing, open for suggestions how to simplify it.

Reply:
Maybe something like this? I'm new to lemmy and rust, so apologies if it is not appropriate for me to comment here. This snippet updates with `default_view` on HTTP 500 as well as HTTP 400, which I assume is desired since that would most likely happen in the OG code anyway on deserialization failure.
```
    let form = client.get(&node_info_url).send()
      .map(|response| {
        response.error_for_status()
          .and_then(|response|  response.json::<NodeInfo>())
          .map_or(default_form, |node_info| {
            InstanceForm::builder()
              .domain(instance.domain)
              .updated(Some(naive_now()))
              .software(node_info.software.and_then(|s| s.name))
              .version(node_info.version.clone())
              .build()
          })
      })?;

      diesel::update(instance::table.find(instance.id))
        .set(form)
        .execute(conn)?;
```

Reply:
Maybe something like this? I'm new to lemmy and rust so apologies if it is not appropriate for me to post this here. It explicitly sets `default_view` on HTTP 500 but I suspect that would happen anyway with the OG code on deserialization failure. Haven't tested this for correctness, but maybe it can be a template for something more readable?

```
    let form_result = client.get(&node_info_url).send()
      .map(|response| {
        response.error_for_status()
          .and_then(|response|  response.json::<NodeInfo>())
          .map_or(default_form, |node_info| {
            InstanceForm::builder()
              .domain(instance.domain)
              .updated(Some(naive_now()))
              .software(node_info.software.and_then(|s| s.name))
              .version(node_info.version.clone())
              .build()
          })
      });

      if let Ok(form) = form_result {
        diesel::update(instance::table.find(instance.id))
          .set(form)
          .execute(conn)?;
      }
```